Checkpoint manager

amtrang opened the pull request.

amtrang committed:
Checkpoint & Manager Draft

Drafted the Checkpoint and Checkpoint Manager scripts. There are a few issues that I need to figure out like how to use a method upon reloading a scene.

amtrang committed:
Checkpoint Manager Draft 2

Still need to fix the error where the Checkpoint Manager is not storing its variables after a scene is reloaded.

amtrang committed:
Checkpoint System Prototype Done

Finished the Checkpoint and Checkpoint Manager prototypes. The player can respawn at a checkpoint, and the manager remembers all of the checkpoints the player passed.

CristionD reviewed (changes requested):
Implement the following modifications:
- Remove the `DontDestroyOnLoad(this.gameObject)` in `MonoSingleton.cs`.
- Attach `CheckpointManager.cs` to the Managers prefab.
- Discard the 'Checkpoint Manager' prefab.
- Insert a serializable array of Transforms into `CheckpointManager`, which shall hold all the Checkpoints in a scene.
- Create an empty game object with a new script that holds the `checkpointsEnabled` bool array, save it as a prefab, add it as a serializable field to `CheckpointManager`, and ensure it is not in the scene before 'Play Mode' commences.
- When a new scene loads, instantiate the game object with the new script (called `CheckpointData.cs` for now) and ensure it is not destroyed upon loading.
    - When a Checkpoint is visited by the Player, update `checkpointsEnabled` in `CheckpointData`.
    - When an existing scene reloads, invoke the `CheckpointManager` to find 'CheckpointData' and acquire `checkpointsEnabled`.
    - Before a new scene loads, destroy `CheckpointData`'s game object.

amtrang committed:
Checkpoint System Revised

Revised the checkpoint system so that the checkpoint manager isn't in DontDestroyOnLoad but uses CheckpointData to store which checkpoints are enabled.

CristionD committed:
Unsubscribe method on disable and remove UserSettings
